File: pkl-core/src/main/java/org/pkl/core/messaging/MessagePackDecoder.java
/**
 * Copyright © 2024 Apple Inc. and the Pkl project authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.pkl.core.messaging;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.file.Path;
import java.time.Duration;
import java.util.*;
import java.util.function.Function;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import org.msgpack.core.MessageTypeException;
import org.msgpack.core.MessageUnpacker;
import org.msgpack.value.Value;
import org.msgpack.value.impl.ImmutableStringValueImpl;
import org.pkl.core.evaluatorSettings.PklEvaluatorSettings;
import org.pkl.core.evaluatorSettings.PklEvaluatorSettings.Proxy;
import org.pkl.core.messaging.Message.*;
import org.pkl.core.module.PathElement;
import org.pkl.core.packages.Checksums;
import org.pkl.core.util.ErrorMessages;
import org.pkl.core.util.Nullable;

public class MessagePackDecoder implements MessageDecoder {

  private final MessageUnpacker unpacker;

  public MessagePackDecoder(MessageUnpacker unpacker) {
    this.unpacker = unpacker;
  }

  @Override
  public @Nullable Message decode() throws IOException, DecodeException {
    if (!unpacker.hasNext()) {
      return null;
    }

    int code;
    try {
      var arraySize = unpacker.unpackArrayHeader();
      if (arraySize != 2) {
        throw new DecodeException(ErrorMessages.create("malformedMessageHeaderLength", arraySize));
      }
      code = unpacker.unpackInt();
    } catch (MessageTypeException e) {
      throw new DecodeException(ErrorMessages.create("malformedMessageHeaderException"), e);
    }

    Type msgType;
    try {
      msgType = Type.fromInt(code);
    } catch (IllegalArgumentException e) {
      throw new DecodeException(
          ErrorMessages.create("malformedMessageHeaderUnrecognizedCode", code), e);
    }

    try {
      var map = unpacker.unpackValue().asMapValue().map();
      return switch (msgType) {
        case CREATE_EVALUATOR_REQUEST ->
            new CreateEvaluatorRequest(
                get(map, "requestId").asIntegerValue().asLong(),
                unpackStringListOrNull(map, "allowedModules", Pattern::compile),
                unpackStringListOrNull(map, "allowedResources", Pattern::compile),
                unpackModuleReaderSpec(map),
                unpackResourceReaderSpec(map),
                unpackStringListOrNull(map, "modulePaths", Path::of),
                unpackStringMapOrNull(map, "env"),
                unpackStringMapOrNull(map, "properties"),
                unpackLongOrNull(map, "timeoutSeconds", Duration::ofSeconds),
                unpackStringOrNull(map, "rootDir", Path::of),
                unpackStringOrNull(map, "cacheDir", Path::of),
                unpackStringOrNull(map, "outputFormat"),
                unpackProject(map),
                unpackHttp(map));
        case CREATE_EVALUATOR_RESPONSE ->
            new CreateEvaluatorResponse(
                unpackLong(map, "requestId"),
                unpackLongOrNull(map, "evaluatorId"),
                unpackStringOrNull(map, "error"));
        case CLOSE_EVALUATOR -> new CloseEvaluator(unpackLong(map, "evaluatorId"));
        case EVALUATE_REQUEST ->
            new EvaluateRequest(
                unpackLong(map, "requestId"),
                unpackLong(map, "evaluatorId"),
                new URI(unpackString(map, "moduleUri")),
                unpackStringOrNull(map, "moduleText"),
                unpackStringOrNull(map, "expr"));
        case EVALUATE_RESPONSE ->
            new EvaluateResponse(
                unpackLong(map, "requestId"),
                unpackLong(map, "evaluatorId"),
                unpackByteArray(map, "result"),
                unpackStringOrNull(map, "error"));
        case LOG_MESSAGE ->
            new LogMessage(
                unpackLong(map, "evaluatorId"),
                unpackInt(map, "level"),
                unpackString(map, "message"),
                unpackString(map, "frameUri"));
        case READ_RESOURCE_REQUEST ->
            new ReadResourceRequest(
                unpackLong(map, "requestId"),
                unpackLong(map, "evaluatorId"),
                new URI(unpackString(map, "uri")));
        case READ_RESOURCE_RESPONSE ->
            new ReadResourceResponse(
                unpackLong(map, "requestId"),
                unpackLong(map, "evaluatorId"),
                unpackByteArray(map, "contents"),
                unpackStringOrNull(map, "error"));
        case READ_MODULE_REQUEST ->
            new ReadModuleRequest(
                unpackLong(map, "requestId"),
                unpackLong(map, "evaluatorId"),
                new URI(unpackString(map, "uri")));
        case READ_MODULE_RESPONSE ->
            new ReadModuleResponse(
                unpackLong(map, "requestId"),
                unpackLong(map, "evaluatorId"),
                unpackStringOrNull(map, "contents"),
                unpackStringOrNull(map, "error"));
        case LIST_RESOURCES_REQUEST ->
            new ListResourcesRequest(
                unpackLong(map, "requestId"),
                unpackLong(map, "evaluatorId"),
                new URI(unpackString(map, "uri")));
        case LIST_RESOURCES_RESPONSE ->
            new ListResourcesResponse(
                unpackLong(map, "requestId"),
                unpackLong(map, "evaluatorId"),
                unpackPathElements(map, "pathElements"),
                unpackStringOrNull(map, "error"));
        case LIST_MODULES_REQUEST ->
            new ListModulesRequest(
                unpackLong(map, "requestId"),
                unpackLong(map, "evaluatorId"),
                new URI(unpackString(map, "uri")));
        case LIST_MODULES_RESPONSE ->
            new ListModulesResponse(
                unpackLong(map, "requestId"),
                unpackLong(map, "evaluatorId"),
                unpackPathElements(map, "pathElements"),
                unpackStringOrNull(map, "error"));
      };

    } catch (MessageTypeException | URISyntaxException e) {
      throw new DecodeException(ErrorMessages.create("malformedMessageBody", code), e);
    }
  }

  private static @Nullable Value getNullable(Map<Value, Value> map, String key) {
    return map.get(new ImmutableStringValueImpl(key));
  }

  private static Value get(Map<Value, Value> map, String key) throws DecodeException {
    var value = map.get(new ImmutableStringValueImpl(key));
    if (value == null) {
      throw new DecodeException(ErrorMessages.create("missingMessageParameter", key));
    }
    return value;
  }

  private static String unpackString(Map<Value, Value> map, String key) throws DecodeException {
    return get(map, key).asStringValue().asString();
  }

  private static <T> T unpackString(Map<Value, Value> map, String key, Function<String, T> mapper)
      throws DecodeException {
    return mapper.apply(get(map, key).asStringValue().asString());
  }

  private static @Nullable String unpackStringOrNull(Map<Value, Value> map, String key) {
    var value = getNullable(map, key);
    if (value == null) {
      return null;
    }
    return value.asStringValue().asString();
  }

  private static <T> @Nullable T unpackStringOrNull(
      Map<Value, Value> map, String key, Function<String, T> mapper) {
    var value = getNullable(map, key);
    if (value == null) {
      return null;
    }
    return mapper.apply(value.asStringValue().asString());
  }

  private static byte[] unpackByteArray(Map<Value, Value> map, String key) {
    var value = getNullable(map, key);
    if (value == null) {
      return new byte[0];
    }
    return value.asBinaryValue().asByteArray();
  }

  private static boolean unpackBoolean(Map<Value, Value> map, String key) throws DecodeException {
    return get(map, key).asBooleanValue().getBoolean();
  }

  private static int unpackInt(Map<Value, Value> map, String key) throws DecodeException {
    return get(map, key).asIntegerValue().asInt();
  }

  private static long unpackLong(Map<Value, Value> map, String key) throws DecodeException {
    return get(map, key).asIntegerValue().asLong();
  }

  private static @Nullable Long unpackLongOrNull(Map<Value, Value> map, String key) {
    var value = getNullable(map, key);
    if (value == null) {
      return null;
    }
    return value.asIntegerValue().asLong();
  }

  private static <T> @Nullable T unpackLongOrNull(
      Map<Value, Value> map, String key, Function<Long, T> mapper) {
    var value = unpackLongOrNull(map, key);
    if (value == null) {
      return null;
    }
    return mapper.apply(value);
  }

  private static @Nullable List<String> unpackStringListOrNull(Map<Value, Value> map, String key) {
    var value = getNullable(map, key);
    if (value == null) {
      return null;
    }

    return value.asArrayValue().list().stream().map((it) -> it.asStringValue().asString()).toList();
  }

  private static @Nullable Map<String, String> unpackStringMapOrNull(
      Map<Value, Value> map, String key) {
    var value = getNullable(map, key);
    if (value == null) {
      return null;
    }

    return value.asMapValue().entrySet().stream()
        .collect(
            Collectors.toMap(
                (e) -> e.getKey().asStringValue().asString(),
                (e) -> e.getValue().asStringValue().asString()));
  }

  private static <T> @Nullable List<T> unpackStringListOrNull(
      Map<Value, Value> map, String key, Function<String, T> mapper) {
    var value = unpackStringListOrNull(map, key);
    if (value == null) {
      return null;
    }

    return value.stream().map(mapper).toList();
  }

  private static @Nullable List<ModuleReaderSpec> unpackModuleReaderSpec(Map<Value, Value> map)
      throws DecodeException {
    var keys = getNullable(map, "clientModuleReaders");
    if (keys == null) {
      return null;
    }

    var result = new ArrayList<ModuleReaderSpec>(keys.asArrayValue().size());
    for (Value value : keys.asArrayValue()) {
      var readerMap = value.asMapValue().map();
      result.add(
          new ModuleReaderSpec(
              unpackString(readerMap, "scheme"),
              unpackBoolean(readerMap, "hasHierarchicalUris"),
              unpackBoolean(readerMap, "isLocal"),
              unpackBoolean(readerMap, "isGlobbable")));
    }
    return result;
  }

  private static @Nullable List<ResourceReaderSpec> unpackResourceReaderSpec(Map<Value, Value> map)
      throws DecodeException {
    var keys = getNullable(map, "clientResourceReaders");
    if (keys == null) {
      return null;
    }

    var result = new ArrayList<ResourceReaderSpec>(keys.asArrayValue().size());
    for (Value value : keys.asArrayValue()) {
      var readerMap = value.asMapValue().map();
      result.add(
          new ResourceReaderSpec(
              unpackString(readerMap, "scheme"),
              unpackBoolean(readerMap, "hasHierarchicalUris"),
              unpackBoolean(readerMap, "isGlobbable")));
    }
    return result;
  }

  private static @Nullable Project unpackProject(Map<Value, Value> map)
      throws DecodeException, URISyntaxException {
    var projVal = getNullable(map, "project");
    if (projVal == null) {
      return null;
    }
    var projMap = projVal.asMapValue().map();
    var projectFileUri = new URI(unpackString(projMap, "projectFileUri"));
    var dependencies = unpackDependencies(projMap, "dependencies");
    return new Project(projectFileUri, null, dependencies);
  }

  private static Map<String, Dependency> unpackDependencies(Map<Value, Value> map, String name)
      throws DecodeException, URISyntaxException {
    var mapValue = get(map, name).asMapValue().map();

    var result = new HashMap<String, Dependency>(mapValue.size());
    for (Map.Entry<Value, Value> entry : mapValue.entrySet()) {
      var dependencyName = entry.getKey().asStringValue().asString();
      var dependencyObj = entry.getValue().asMapValue().map();
      var type = unpackString(dependencyObj, "type");
      var packageUri = new URI(unpackString(dependencyObj, "packageUri"));

      if (type.equals(Dependency.Type.REMOTE.getType())) {
        var checksumsObj = getNullable(dependencyObj, "checksums");
        if (checksumsObj == null) {
          result.put(dependencyName, new RemoteDependency(packageUri, null));
        } else {
          result.put(
              dependencyName,
              new RemoteDependency(
                  packageUri,
                  new Checksums(unpackString(checksumsObj.asMapValue().map(), "sha256"))));
        }
      } else {
        var dependencies = unpackDependencies(dependencyObj, "dependencies");
        var projectFileUri = unpackString(dependencyObj, "projectFileUri");
        result.put(dependencyName, new Project(new URI(projectFileUri), packageUri, dependencies));
      }
    }
    return result;
  }

  private static @Nullable Http unpackHttp(Map<Value, Value> map) throws DecodeException {
    var httpVal = getNullable(map, "http");
    if (httpVal == null) {
      return null;
    }

    var httpMap = httpVal.asMapValue().map();
    var proxy = unpackProxy(httpMap);

    var caCertificates = getNullable(httpMap, "caCertificates");
    if (caCertificates == null) {
      return new Http(new byte[0], proxy);
    }

    return new Http(caCertificates.asBinaryValue().asByteArray(), proxy);
  }

  private static @Nullable PklEvaluatorSettings.Proxy unpackProxy(Map<Value, Value> map)
      throws DecodeException {
    var proxyVal = getNullable(map, "proxy");
    if (proxyVal == null) {
      return null;
    }
    var proxyMap = proxyVal.asMapValue().map();
    var address = unpackString(proxyMap, "address");
    var noProxy = unpackStringListOrNull(proxyMap, "noProxy");
    return Proxy.create(address, noProxy);
  }

  private static @Nullable List<PathElement> unpackPathElements(Map<Value, Value> map, String key)
      throws DecodeException {
    var value = getNullable(map, key);
    if (value == null) {
      return null;
    }

    var result = new ArrayList<PathElement>(value.asArrayValue().size());
    for (Value pathElement : value.asArrayValue()) {
      var pathElementMap = pathElement.asMapValue().map();
      result.add(
          new PathElement(
              unpackString(pathElementMap, "name"), unpackBoolean(pathElementMap, "isDirectory")));
    }
    return result;
  }
}

File: pkl-server/src/main/kotlin/org/pkl/server/MessagePackDecoder.kt
/**
 * Copyright © 2024 Apple Inc. and the Pkl project authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.pkl.server

import java.net.URI
import java.nio.file.Path
import java.time.Duration
import java.util.regex.Pattern
import org.msgpack.core.MessageTypeException
import org.msgpack.core.MessageUnpacker
import org.msgpack.value.Value
import org.msgpack.value.impl.ImmutableStringValueImpl
import org.pkl.core.evaluatorSettings.PklEvaluatorSettings
import org.pkl.core.module.PathElement
import org.pkl.core.packages.Checksums

internal class MessagePackDecoder(private val unpacker: MessageUnpacker) : MessageDecoder {
  override fun decode(): Message? {
    if (!unpacker.hasNext()) return null

    val code =
      try {
        val arraySize = unpacker.unpackArrayHeader()
        if (arraySize != 2) {
          throw DecodeException("Malformed message header (expected size 2, but got $arraySize).")
        }
        unpacker.unpackInt()
      } catch (e: MessageTypeException) {
        throw DecodeException("Malformed message header.", e)
      }

    return try {
      val map = unpacker.unpackValue().asMapValue().map()
      when (code) {
        MessageType.CREATE_EVALUATOR_REQUEST.code -> {
          CreateEvaluatorRequest(
            requestId = map.get("requestId").asIntegerValue().asLong(),
            allowedModules = map.unpackStringListOrNull("allowedModules")?.map(Pattern::compile),
            allowedResources =
              map.unpackStringListOrNull("allowedResources")?.map(Pattern::compile),
            clientModuleReaders = map.unpackModuleReaderSpec("clientModuleReaders"),
            clientResourceReaders = map.unpackResourceReaderSpec("clientResourceReaders"),
            modulePaths = map.unpackStringListOrNull("modulePaths")?.map(Path::of),
            env = map.unpackStringMapOrNull("env"),
            properties = map.unpackStringMapOrNull("properties"),
            timeout = map.unpackLongOrNull("timeoutSeconds")?.let(Duration::ofSeconds),
            rootDir = map.unpackStringOrNull("rootDir")?.let(Path::of),
            cacheDir = map.unpackStringOrNull("cacheDir")?.let(Path::of),
            outputFormat = map.unpackStringOrNull("outputFormat"),
            project = map.unpackProject("project"),
            http = map.unpackHttp("http"),
          )
        }
        MessageType.CREATE_EVALUATOR_RESPONSE.code -> {
          CreateEvaluatorResponse(
            requestId = map.unpackLong("requestId"),
            evaluatorId = map.unpackLongOrNull("evaluatorId"),
            error = map.unpackStringOrNull("error")
          )
        }
        MessageType.CLOSE_EVALUATOR.code -> {
          CloseEvaluator(evaluatorId = map.unpackLong("evaluatorId"))
        }
        MessageType.EVALUATE_REQUEST.code -> {
          EvaluateRequest(
            requestId = map.unpackLong("requestId"),
            evaluatorId = map.unpackLong("evaluatorId"),
            moduleUri = map.unpackString("moduleUri").let(::URI),
            moduleText = map.unpackStringOrNull("moduleText"),
            expr = map.unpackStringOrNull("expr")
          )
        }
        MessageType.EVALUATE_RESPONSE.code -> {
          EvaluateResponse(
            requestId = map.unpackLong("requestId"),
            evaluatorId = map.unpackLong("evaluatorId"),
            result = map.unpackByteArrayOrNull("result"),
            error = map.unpackStringOrNull("error")
          )
        }
        MessageType.LOG_MESSAGE.code -> {
          LogMessage(
            evaluatorId = map.unpackLong("evaluatorId"),
            level = map.unpackIntValue("level"),
            message = map.unpackString("message"),
            frameUri = map.unpackString("frameUri")
          )
        }
        MessageType.READ_RESOURCE_REQUEST.code -> {
          ReadResourceRequest(
            requestId = map.unpackLong("requestId"),
            evaluatorId = map.unpackLong("evaluatorId"),
            uri = map.unpackString("uri").let(::URI)
          )
        }
        MessageType.READ_RESOURCE_RESPONSE.code -> {
          ReadResourceResponse(
            requestId = map.unpackLong("requestId"),
            evaluatorId = map.unpackLong("evaluatorId"),
            contents = map.unpackByteArrayOrNull("contents"),
            error = map.unpackStringOrNull("error")
          )
        }
        MessageType.READ_MODULE_REQUEST.code -> {
          ReadModuleRequest(
            requestId = map.unpackLong("requestId"),
            evaluatorId = map.unpackLong("evaluatorId"),
            uri = map.unpackString("uri").let(::URI)
          )
        }
        MessageType.READ_MODULE_RESPONSE.code -> {
          ReadModuleResponse(
            requestId = map.unpackLong("requestId"),
            evaluatorId = map.unpackLong("evaluatorId"),
            contents = map.unpackStringOrNull("contents"),
            error = map.unpackStringOrNull("error")
          )
        }
        MessageType.LIST_MODULES_REQUEST.code -> {
          ListModulesRequest(
            requestId = map.unpackLong("requestId"),
            evaluatorId = map.unpackLong("evaluatorId"),
            uri = map.unpackString("uri").let(::URI)
          )
        }
        MessageType.LIST_MODULES_RESPONSE.code -> {
          ListModulesResponse(
            requestId = map.unpackLong("requestId"),
            evaluatorId = map.unpackLong("evaluatorId"),
            pathElements = map.unpackPathElements("pathElements"),
            error = map.unpackStringOrNull("error")
          )
        }
        MessageType.LIST_RESOURCES_REQUEST.code -> {
          ListResourcesRequest(
            requestId = map.unpackLong("requestId"),
            evaluatorId = map.unpackLong("evaluatorId"),
            uri = map.unpackString("uri").let(::URI)
          )
        }
        MessageType.LIST_RESOURCES_RESPONSE.code -> {
          ListResourcesResponse(
            requestId = map.unpackLong("requestId"),
            evaluatorId = map.unpackLong("evaluatorId"),
            pathElements = map.unpackPathElements("pathElements"),
            error = map.unpackStringOrNull("error")
          )
        }
        else -> throw ProtocolException("Invalid message code: $code")
      }
    } catch (e: MessageTypeException) {
      throw DecodeException("Malformed message body for message with code `$code`.", e)
    }
  }

  private fun Array<Value>.unpackValueOrNull(key: String): Value? {
    for (i in indices.step(2)) {
      val currKey = this[i].asStringValue().asString()
      if (currKey == key) return this[i + 1]
    }
    return null
  }

  private fun Map<Value, Value>.getNullable(key: String): Value? =
    this[ImmutableStringValueImpl(key)]

  private fun Map<Value, Value>.get(key: String): Value =
    getNullable(key) ?: throw DecodeException("Missing message parameter `$key`")

  private fun Array<Value>.unpackValue(key: String): Value =
    unpackValueOrNull(key) ?: throw DecodeException("Missing message parameter `$key`.")

  private fun Map<Value, Value>.unpackStringListOrNull(key: String): List<String>? {
    val value = getNullable(key) ?: return null
    return value.asArrayValue().map { it.asStringValue().asString() }
  }

  private fun Map<Value, Value>.unpackStringMapOrNull(key: String): Map<String, String>? {
    val value = getNullable(key) ?: return null
    return value.asMapValue().entrySet().associate { (k, v) ->
      k.asStringValue().asString() to v.asStringValue().asString()
    }
  }

  private fun Map<Value, Value>.unpackLong(key: String): Long = get(key).asIntegerValue().asLong()

  private fun Map<Value, Value>.unpackBoolean(key: String): Boolean =
    get(key).asBooleanValue().boolean

  private fun Map<Value, Value>.unpackBooleanOrNull(key: String): Boolean? =
    getNullable(key)?.asBooleanValue()?.boolean

  private fun Map<Value, Value>.unpackLongOrNull(key: String): Long? =
    getNullable(key)?.asIntegerValue()?.asLong()

  private fun Map<Value, Value>.unpackIntValue(key: String): Int = get(key).asIntegerValue().asInt()

  private fun Map<Value, Value>.unpackString(key: String): String =
    get(key).asStringValue().asString()

  private fun Map<Value, Value>.unpackStringOrNull(key: String): String? =
    getNullable(key)?.asStringValue()?.asString()

  private fun Map<Value, Value>.unpackByteArrayOrNull(key: String): ByteArray? =
    getNullable(key)?.asBinaryValue()?.asByteArray()

  private fun Map<Value, Value>.unpackPathElements(key: String): List<PathElement>? =
    getNullable(key)?.asArrayValue()?.map { pathElement ->
      val map = pathElement.asMapValue().map()
      PathElement(map.unpackString("name"), map.unpackBoolean("isDirectory"))
    }

  private fun Map<Value, Value>.unpackModuleReaderSpec(name: String): List<ModuleReaderSpec>? {
    val keys = getNullable(name) ?: return null
    return keys.asArrayValue().toList().map { value ->
      val readerMap = value.asMapValue().map()
      ModuleReaderSpec(
        scheme = readerMap.unpackString("scheme"),
        hasHierarchicalUris = readerMap.unpackBoolean("hasHierarchicalUris"),
        isLocal = readerMap.unpackBoolean("isLocal"),
        isGlobbable = readerMap.unpackBoolean("isGlobbable")
      )
    }
  }

  private fun Map<Value, Value>.unpackResourceReaderSpec(name: String): List<ResourceReaderSpec> {
    val keys = getNullable(name) ?: return emptyList()
    return keys.asArrayValue().toList().map { value ->
      val readerMap = value.asMapValue().map()
      ResourceReaderSpec(
        scheme = readerMap.unpackString("scheme"),
        hasHierarchicalUris = readerMap.unpackBoolean("hasHierarchicalUris"),
        isGlobbable = readerMap.unpackBoolean("isGlobbable")
      )
    }
  }

  private fun Map<Value, Value>.unpackProject(name: String): Project? {
    val projMap = getNullable(name)?.asMapValue()?.map() ?: return null
    val projectFileUri = URI(projMap.unpackString("projectFileUri"))
    val dependencies = projMap.unpackDependencies("dependencies")
    return Project(projectFileUri, null, dependencies)
  }

  private fun Map<Value, Value>.unpackHttp(name: String): PklEvaluatorSettings.Http? {
    val httpMap = getNullable(name)?.asMapValue()?.map() ?: return null
    val proxy = httpMap.unpackProxy("proxy")
    return PklEvaluatorSettings.Http(proxy)
  }

  private fun Map<Value, Value>.unpackProxy(name: String): PklEvaluatorSettings.Proxy? {
    val proxyMap = getNullable(name)?.asMapValue()?.map() ?: return null
    val address = proxyMap.unpackString("address")
    val noProxy = proxyMap.unpackStringListOrNull("noProxy")
    return PklEvaluatorSettings.Proxy.create(address, noProxy)
  }

  private fun Map<Value, Value>.unpackDependencies(name: String): Map<String, Dependency> {
    val mapValue = get(name).asMapValue().map()
    return mapValue.entries.associate { (key, value) ->
      val dependencyName = key.asStringValue().asString()
      val dependencyObj = value.asMapValue().map()
      val type = dependencyObj.unpackString("type")
      val packageUri = URI(dependencyObj.unpackString("packageUri"))
      if (type == DependencyType.REMOTE.value) {
        val checksums =
          dependencyObj.getNullable("checksums")?.asMapValue()?.map()?.let { obj ->
            val sha256 = obj.unpackString("sha256")
            Checksums(sha256)
          }
        return@associate dependencyName to RemoteDependency(packageUri, checksums)
      }
      val dependencies = dependencyObj.unpackDependencies("dependencies")
      val projectFileUri = dependencyObj.unpackString("projectFileUri")
      dependencyName to Project(URI(projectFileUri), packageUri, dependencies)
    }
  }
}
